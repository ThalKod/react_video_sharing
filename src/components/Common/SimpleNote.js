import React from "react";

export default function SimpleNote({ text}){
  return (
      <div className="content-wrapper">
        <div className="container">
          <div className="row">
            <div className="col-lg-12 upload-page">
              <div className="u-area">
                {text}
              </div>
            </div>
          </div>
        </div>
      </div>
      )
};
