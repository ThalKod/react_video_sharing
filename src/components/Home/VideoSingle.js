import React from "react";
import { Link } from "react-router-dom";

import { request, formatSecondForRendering } from "utils";

// import videoPlaceholder from "../assets/images/video_placeholder.png";

class VideoSingle extends React.Component{

  state = {
    videoImage: null,
    // videoThumbnails: null,
  };

  componentDidMount = () => {
    const { _id } = this.props;

    request("get", `/video/cover/default/${_id}`)
        .then(res => {
          if(!res.data.error) this.setState({ videoImage: res.data.coverPhoto })
        })
        .catch(err => console.log(err));
  };

  render(){
    const { name, duration, viewCount, _id } = this.props;
    const { videoImage } = this.state;

    return (
        <div className="col-lg-3 col-sm-6 videoitem">
          <div className="b-video">
            <div className="v-img">
              <Link to={`/video/${_id}`}><img src={videoImage} alt="videoimage" /></Link>
              <div className="time">{formatSecondForRendering(duration)}</div>
            </div>
            <div className="v-desc">
              <Link to={`/video/${_id}`}>{name}</Link>
            </div>
            <div className="v-views">
              {viewCount} views.
            </div>
          </div>
        </div>
    )
  }
}

export default VideoSingle;

