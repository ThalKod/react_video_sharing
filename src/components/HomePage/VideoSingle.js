import React from "react";

import { request, formatSecondForRendering } from "../../utils/index";

// import videoPlaceholder from "../assets/images/video_placeholder.png";

class VideoSingle extends React.Component{

  state = {
    videoImage: null,
    // videoThumbnails: null,
  };

  componentDidMount = () => {
    const { _id } = this.props;

    request("get", `/video/cover/default/${_id}`)
        .then(res => {
          if(!res.data.error) this.setState({ videoImage: res.data.coverPhoto })
        })
        .catch(err => console.log(err));
  };

  render(){
    const { name, duration, viewCount } = this.props;
    const { videoImage } = this.state;

    return (
        <div className="col-lg-3 col-sm-6 videoitem">
          <div className="b-video">
            <div className="v-img">
              <a href="/"><img src={videoImage} alt="placeholder" /></a>
              <div className="time">{formatSecondForRendering(duration)}</div>
            </div>
            <div className="v-desc">
              <a href="/">{name}</a>
            </div>
            <div className="v-views">
              {viewCount} views.
            </div>
          </div>
        </div>
    )
  }
}

export default VideoSingle;

