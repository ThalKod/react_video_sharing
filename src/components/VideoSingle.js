import React from "react";

import videoPlaceholder from "../assets/images/video_placeholder.png";

export default ({ name, duration, viewsCount }) => {
  return (
      <div className="col-lg-3 col-sm-6 videoitem">
        <div className="b-video">
          <div className="v-img">
            <a href="/"><img src={videoPlaceholder} alt="placeholder" /></a>
            <div className="time">{duration}</div>
          </div>
          <div className="v-desc">
            <a href="/">{name}</a>
          </div>
          <div className="v-views">
            {viewsCount} views.
          </div>
        </div>
      </div>
  )
}
