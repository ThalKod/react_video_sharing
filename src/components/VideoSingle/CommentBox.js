import React from "react";
import { connect } from "react-redux";

import  { startAddComments, startAddReply } from "actions";

class CommentBox extends React.Component {

  state = {
    commentText: ""
  };


  handleCommentSubmit = (e) => {
    e.preventDefault();

    const { commentText  } = this.state;
    const { id, addComments, onSubmit, reply, addReply, commentId } = this.props;

    if(!commentText) return;

    if(!reply){
       addComments({ commentText } , id)
          .then(({ error }) =>{
            if(!error)
              this.setState({ commentText: "" });
            onSubmit();
            // then handling error
          })
          .catch(err => console.log(err));
    }else{
      addReply({ commentText }, commentId)
          .then(({ error }) => {
            if(!error)
              this.setState({ commentText: "" });
            onSubmit();
          })
          .catch(err => console.log(err));
    }
  };

  render() {

    const { commentText } = this.state;
    const { reply, hideReplyBox } = this.props;

    return (
        <div className="reply-comment">
          <div className="rc-comment">
            <form>
                      <textarea
                          value={commentText}
                          // eslint-disable-next-line jsx-a11y/no-autofocus
                          autoFocus={!!reply}
                          onChange={({ target: { value }}) => { this.setState({ commentText: value }) }}
                          rows="3"
                          placeholder="Enter comment here..."
                          onBlur={reply ? hideReplyBox : null }
                      />
              <button type="button" onMouseDown={this.handleCommentSubmit}>
                <i className="cv cvicon-cv-add-comment"/>
              </button>
            </form>
          </div>
        </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  addComments: (comment, id) => dispatch(startAddComments(comment, id)),
  addReply: (comment, id) => dispatch(startAddReply(comment, id)),
});

export default connect(null, mapDispatchToProps)(CommentBox);
