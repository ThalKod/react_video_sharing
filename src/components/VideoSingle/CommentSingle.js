import React from "react";
import { Link } from "react-router-dom"

import Avatar from "components/Header/Avatar";
import { formatTimestamps } from "utils";


export default class CommentSingle extends React.Component{

  state= {
    replyBox: false
  };

  reply = () => {
    this.setState({ replyBox: true});
  };

  render(){
    const { text, author: { username, _id }, createdAt } = this.props;
    const { replyBox } = this.state;

    return(
        <div className="cl-comment">
          <div className="cl-avatar">
            <Link to={`/channel/${_id}`}>
              <Avatar username={username}/>
            </Link>
          </div>
          <div className="cl-comment-text">
            <div className="cl-name-date"><Link to={`/channel/${_id}`}>{username}</Link> . {formatTimestamps(createdAt)}</div>
            <div className="cl-text">{text}</div>
            { !replyBox ?
                <div className="cl-meta">
                  <button onClick={this.reply} className="reply-button" type="submit">Reply</button>
                </div>
                :
                <div className="reply-comment">
                  <div className="rc-comment">
                    <form>
                      <textarea
                          rows="3"
                          placeholder="Enter comment here..."
                      />
                      <button type="submit" onClick={this.handleCommentSubmit}>
                        <i className="cv cvicon-cv-add-comment"/>
                      </button>
                    </form>
                  </div>
                </div>
            }
          </div>
          <div className="clearfix"/>
        </div>
    )
  }
}

