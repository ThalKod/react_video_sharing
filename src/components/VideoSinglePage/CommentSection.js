import React from "react";
import { connect } from "react-redux";

import Avatar from "../Header/Avatar";
import { request } from "../../utils";

class CommentSection extends React.Component{

  state = {
    commentText: "Enter your comment here...",
  };


  checkIfCanComment = () => {
    const  { userName, redirect } = this.props;
    if(!userName) redirect();
  };

  handleCommentSubmit = (e) => {
    e.preventDefault();

    const { commentText } = this.state;

    request("post", "/comments/new", {}, { commentText })
        .then(res => console.log(res))
        .catch(err => console.log(err));
  };

  render(){

    const { userName } = this.props;
    const { commentText} = this.state;

    return(
        <div className="comments">
          <div className="reply-comment">
            <div className="rc-header"><i className="cv cvicon-cv-comment"/> <span
                className="semibold">236</span> Comments
            </div>
            { userName && <div className="rc-ava">
                            <Avatar username={userName}/>
                          </div>
            }
            <div className="rc-comment">
              <form>
                <textarea
                    rows="3"
                    value={commentText}
                    onChange={({ target: { value }}) => { this.setState({ commentText: value }) }}
                    onFocus={this.checkIfCanComment}
                    />
                <button type="submit" onClick={this.handleCommentSubmit}>
                  <i className="cv cvicon-cv-add-comment"/>
                </button>
              </form>
            </div>
            <div className="clearfix"/>
          </div>
        </div>
    )
  }
}

const mapStateToProps = (state) => ({
  userName: state.user.username,
});

export default connect(mapStateToProps)(CommentSection);
