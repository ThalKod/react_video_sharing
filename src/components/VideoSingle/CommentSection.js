import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

import Avatar from "components/Header/Avatar";
import CommentSingle from "components/VideoSingle/CommentSingle";
import { request } from "utils";
import { startGetComments, startAddComments, clearComment } from "actions";
import SimpleNote from "components/Common/SimpleNote";

export class CommentSection extends React.Component{

  state = {
    commentText: "",
    commentTotal: 0,
    loadMore: false
  };

  checkIfCanComment = () => {
    const  { userName, redirect } = this.props;
    if(!userName) redirect();
  };

  handleCommentSubmit = (e) => {
    e.preventDefault();

    const { commentText, commentTotal } = this.state;
    const { id, addComments } = this.props;

    if(!commentText) return;

    addComments({ commentText } , id)
        .then(({ error }) =>{
          if(!error)
            this.setState({ commentText: "", commentTotal: commentTotal + 1 });
          // then handling error
        })
        .catch(err => console.log(err));
  };

  loadMoreComment = () => {
    const { getComments, offset, id } = this.props;

    // TODO: Handling loading and error
    getComments({ offset }, id)
        .then(({ error, end}) => {
          if(!error) {
            if(end) this.setState({ loadMore: false });
          }
        })
        .catch(err => console.log(err));
  };

  renderCommentList = () => {
    const { comments } = this.props;
    if(!comments || comments.length <= 0) return <SimpleNote medium text="No Comments"/>;

    return comments.map(({ _id, text, author, createdAt}) => {
      return <CommentSingle text={text} key={_id} author={author} createdAt={createdAt}/>
    });
  };

  componentDidMount = () => {
    const { id, getComments, offset } = this.props;

    const getCommentTotalPromise = request("get",  `/comment/count/video/${id}`);
    const getCommentsListPromise = getComments({ offset }, id);

    Promise.all([getCommentTotalPromise, getCommentsListPromise])
        .then(res => {
          if(!res[0].data.error && !res[1].error){
            const { count } = res[0].data;
            this.setState({ commentTotal: count, loadMore: count > 0 });
          }

        })
        .catch(err => console.log(err));
  };

  componentWillUnmount = () => {
    const { clearComments } = this.props;
    clearComments();
  };

  render(){

    const { userName, userId } = this.props;
    const { commentText, commentTotal, loadMore } = this.state;

    return(
        <div className="comments">
          <div className="reply-comment">
            <div className="rc-header"><i className="cv cvicon-cv-comment"/> <span
                className="semibold">{commentTotal}</span> Comments
            </div>
            { userName && <div className="rc-ava">
                            <Link to={`/channel/${userId}`}>
                              <Avatar username={userName}/>
                            </Link>
                          </div>
            }
            <div className="rc-comment">
              <form>
                <textarea
                    rows="3"
                    value={commentText}
                    placeholder="Enter comment here..."
                    onChange={({ target: { value }}) => { this.setState({ commentText: value }) }}
                    onFocus={this.checkIfCanComment}
                    />
                <button type="submit" onClick={this.handleCommentSubmit}>
                  <i className="cv cvicon-cv-add-comment"/>
                </button>
              </form>
            </div>
            <div className="clearfix"/>
          </div>
          <div className="comments-list">
            <div className="cl-header">
              <div className="c-nav">
                <ul className="list-inline">
                  <li className="active">Newest Comments</li>
                </ul>
              </div>
            </div>
            {this.renderCommentList()}
            <div className="row">
              <div className="col-lg-12">
                <div className="loadmore-comments">
                  {loadMore && <button onClick={this.loadMoreComment} type="submit" className="btn btn-default h-btn">Load more Comments</button>}
                </div>
              </div>
            </div>
          </div>
        </div>
    )
  }
}

const mapStateToProps = (state) => ({
  userName: state.user.username,
  userId: state.user._id,
  comments: state.comment.comments,
  offset: state.comment.offset,
});

const mapDispatchToProps = (dispatch) => ({
  getComments: (options, id) => dispatch(startGetComments(options, id)),
  addComments: (comment, id) => dispatch(startAddComments(comment, id)),
  clearComments: () => dispatch(clearComment())
});

export default connect(mapStateToProps, mapDispatchToProps)(CommentSection);
