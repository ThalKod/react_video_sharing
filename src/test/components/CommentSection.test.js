import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { CommentSection } from "components/VideoSingle/CommentSection";
import SimpleNote from "components/Common/SimpleNote";

describe("CommentSection", () => {

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<CommentSection getComments={() => {}}/>);
  });

  it("should render the components correctly if no comments", () => {
    expect(
        wrapper.containsMatchingElement(
            <SimpleNote
                medium
                text="No Comments"
            />
        )
    ).toBe(true);
  });

  it("should render a list of comments", () => {

    const comments = [
      {
        _id: "dLKasdsN1212i12LNK",
        text: "Hello World !",
        author: { username: "Joseph" },
        createdAt: 1129910012010,
      },
      {
        _id: "dLKN1212sdfsi12LNK",
        text: "Hello World !",
        author: { username: "Tim" },
        createdAt: 1129910012010,
      },
      {
        _id: "dLKN12sdsd12i12LNK",
        text: "Hello World !",
        author: { username: "Sarah" },
        createdAt: 1129910012010,
      }
    ];

    wrapper.setProps({ comments });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
