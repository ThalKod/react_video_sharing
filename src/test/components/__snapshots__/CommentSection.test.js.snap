// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentSection should render a list of comments 1`] = `
<div
  className="comments"
>
  <div
    className="reply-comment"
  >
    <div
      className="rc-header"
    >
      <i
        className="cv cvicon-cv-comment"
      />
       
      <span
        className="semibold"
      >
        0
      </span>
       Comments
    </div>
    <Connect(CommentBox)
      onSubmit={[Function]}
    />
    <div
      className="clearfix"
    />
  </div>
  <div
    className="comments-list"
  >
    <div
      className="cl-header"
    >
      <div
        className="c-nav"
      >
        <ul
          className="list-inline"
        >
          <li
            className="active"
          >
            Newest Comments
          </li>
        </ul>
      </div>
    </div>
    <CommentSingle
      author={
        Object {
          "username": "Joseph",
        }
      }
      createdAt={1129910012010}
      id="dLKasdsN1212i12LNK"
      key="dLKasdsN1212i12LNK"
      text="Hello World !"
    />
    <CommentSingle
      author={
        Object {
          "username": "Tim",
        }
      }
      createdAt={1129910012010}
      id="dLKN1212sdfsi12LNK"
      key="dLKN1212sdfsi12LNK"
      text="Hello World !"
    />
    <CommentSingle
      author={
        Object {
          "username": "Sarah",
        }
      }
      createdAt={1129910012010}
      id="dLKN12sdsd12i12LNK"
      key="dLKN12sdsd12i12LNK"
      text="Hello World !"
    />
    <div
      className="row"
    >
      <div
        className="col-lg-12"
      >
        <div
          className="loadmore-comments"
        />
      </div>
    </div>
  </div>
</div>
`;
